ksqldb-engine/src/main/java/io/confluent/ksql/engine/RuntimeAssignor.java

```diff
@@ -57,11 +63,28 @@ public RuntimeAssignor createSandbox() {
     return new RuntimeAssignor(this);
   }
 
-  public String getRuntimeAndMaybeAddRuntime(final QueryId queryId,
-                                             final Collection<SourceName> sources,
-                                             final KsqlConfig config) {
-    if (idToRuntime.containsKey(queryId)) {
-      return idToRuntime.get(queryId);
+  synchronized public String getRuntimeAndMaybeAddRuntime(final QueryId queryId,
```

Comment:
Why only synchronize this function but not others like `dropQuery` and `rebuildAssignment` ? It seems both `runtimesToSources` and `idToRuntime` are updated there too.